### ISSUE TITLE
Simsearch query segfault when using IVF indexes, but not default/flat index

### ISSUE THREAD

**Issue by dgtlmoon** [jolibrain/deepdetect#1368]
#### Configuration

- Version of DeepDetect:
    - [ ] Locally compiled on:
        - [ ] Ubuntu 18.04 LTS
        - [ ] Other:
    - [x] Docker CPU
    - [x] Docker GPU
    - [ ] Amazon AMI
- Commit heads/v0.19.0:1673a99ecc922e01dd7cc8845098291ef46a8902

#### Your question / the problem you're facing:

Performing a prediction/search using simsearch, after building the index, but ONLY when using IVF style indexes causes DD to segfault and crash, in other words ;

- Building the index works fine (6000 images) `IVF20,SQ8`
- Train/Writing the index to disk works fine
- When the search is performed, it crashes

When I try without the extra options, it works fine, but it's far too slow for predict/query search lookups (0.8-0.9s per 150,000 images)
```
'index': True, "ondisk": True,
> remove "index_type": "IVF20,SQ8",
> remove "train_samples": 100,
> remove "nprobe": 64
```

I'm using my finetuned classifier `se_resnet_50`

### Create a search service after indexing and query

What's interesting here, is that it will crash before it even gets a chance to write the log, there is no log here that it received the prediction call.. maybe this is a clue, here you can see the `PUT` but no `POST`

```
[2021-11-14 21:29:30.315] [api] [info] HTTP/1.1 "PUT /services/test" <n/a> 201 551ms
open existing index db
[2021-11-14 21:30:08.347] [torchlib] [info] Opened lmdb /var/www/xxx/web/files/trainer/simsearch/model//names.bin
bash: line 1:     7 Segmentation fault      (core dumped) ./dede -host 0.0.0.0
```

the setup..

```
curl -X PUT "http://localhost:8080/services/test" -d '{
       "mllib":"caffe",
       "description":"similarity search service",
       "type":"unsupervised",
       "parameters":{
         "input":{
           "connector":"image",
           "height": 224,
           "width": 224
         },
         "mllib":{
           "nclasses":20
         }
       },
       "model":{
          "templates":"../templates/caffe/",
          "repository":"/var/www/xxx/web/files/trainer/simsearch/model/",
          "weight": "model_iter_13500.caffemodel"
       }
     }'

sleep 1
curl -X POST "http://localhost:8080/predict" -d '{
       "service":"test",
       "parameters":{
         "input":{ "height": 224, "width": 224  },
         "output":{ "search_nn": 10, "search": true  },
         "mllib":{ "extract_layer":"pool5/7x7_s1"  }
       },
       "data":["https://www.google.com/images/branding/googlelogo/1x/googlelogo_color_272x92dp.png"]  }'
```

### Indexing Service setup

```
curl -X PUT "http://localhost:8080/services/simsearch" -d '{
       "mllib":"caffe",
       "description":"similarity search service",
       "type":"unsupervised",
       "parameters":{
         "input":{
           "connector":"image",
           "height": 224,
           "width": 224
         },
         "mllib":{
           "nclasses": 20
         }
       },
       "model":{
         "repository":"/var/www/xxx/web/files/trainer/simsearch/model/",
         "templates":"/opt/deepdetect/build/templates/caffe"
       }
     }'
```

### Indexing Index call

```
    r = requests.post('http://127.0.0.1:8080/predict', json={"service": "simsearch",
                                                             'parameters': {
                                                                 'input': {'width': 224, 'height': 224},
                                                                 'output': {'index': True, "ondisk": True,
                                                                            "index_type": "IVF20,SQ8",
                                                                            "train_samples": 100,
                                                                            "nprobe": 64
                                                                            },
                                                                 'mllib': {'extract_layer': "pool5/7x7_s1"},
                                                             },
                                                             'data': urls
                                                             })
```

### Indexing Writing the index

```
    r = requests.post('http://127.0.0.1:8080/predict', json={"service": "simsearch",
                                                             'parameters': {
                                                                 'input': {'width': 224, 'height': 224},
                                                                 'output': {'index': False, 'build_index': True},
                                                                 'mllib': {'extract_layer': "pool5/7x7_s1"},
                                                             },
                                                             'data': [existing_uri]
                                                             })
```

The `model/` dir looks all good

```
trainer/simsearch$ ls -al model
-rw-r--r-- 1 dgtlmoon dgtlmoon        28 Nov  7 16:47 best_model.txt
-rw-r--r-- 1 dgtlmoon dgtlmoon       112 Nov  7 16:47 corresp.txt
-rw-rw-r-- 1 dgtlmoon dgtlmoon     16384 Nov 11 16:10 database
-rw-r--r-- 1 dgtlmoon dgtlmoon     51084 Nov 14 22:29 deploy.prototxt
-rw-rw-r-- 1 dgtlmoon dgtlmoon       272 Nov 11 16:19 f.py
-rw-r--r-- 1 dgtlmoon dgtlmoon    183484 Nov 14 22:16 index.faiss
-rw-r--r-- 1 dgtlmoon dgtlmoon  67108864 Nov 14 22:16 index_mmap.faiss
-rw-r--r-- 1 dgtlmoon dgtlmoon 104570495 Nov  7 16:46 model_iter_13500.caffemodel
-rw-r--r-- 1 dgtlmoon dgtlmoon       524 Nov 14 22:29 model.json
drwxr--r-- 2 dgtlmoon dgtlmoon      4096 Nov 14 20:12 names.bin
-rw-r--r-- 1 dgtlmoon dgtlmoon     53024 Nov  7 16:47 se_resnet_50.prototxt
-rw-r--r-- 1 dgtlmoon dgtlmoon       298 Nov  7 16:47 se_resnet_50_solver.prototxt
```

**Comment by dgtlmoon** [comment]
I tried 600 images, with a much lower `train_samples` and `nprobe` but the same problem..

```
    "index_type": "IVF20,SQ8",
    "train_samples": 10,
    "nprobe": 8

```

**Comment by beniz** [comment]
First investigation reveals this is faiss crashing: 
```
Thread 37 "dede" received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x7ffe50ffd700 (LWP 10616)]
0x0000555555c49c41 in faiss::OnDiskInvertedLists::OngoingPrefetch::Thread::one_list() ()
(gdb) bt
#0  0x0000555555c49c41 in faiss::OnDiskInvertedLists::OngoingPrefetch::Thread::one_list() ()
#1  0x0000555555c49d7b in faiss::OnDiskInvertedLists::OngoingPrefetch::prefetch_list(void*) ()
#2  0x00007fffda3366db in start_thread (arg=0x7ffe50ffd700) at pthread_create.c:463
#3  0x00007fffd20aa71f in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95
```

**Comment by dgtlmoon** [comment]
Maybe the version of FAISS changed, and it worked with an earlier lib version?

**Comment by dgtlmoon** [comment]
The cmake doesnt seem to be tied to any particular version

```
      message(STATUS "Fetching FAISS")
      add_definitions(-DUSE_FAISS)
      add_definitions(-DUSE_SIMSEARCH)
      ExternalProject_Add(
        faisslib
        PREFIX faiss
        SOURCE_DIR ${CMAKE_BINARY_DIR}/faiss/src/faiss
        GIT_REPOSITORY https://github.com/facebookresearch/faiss
        GIT_TAG main
        UPDATE_DISCONNECTED 1
        CONFIGURE_COMMAND cmake ${CONFIGURE_OPTS} -B build .
        BUILD_COMMAND make -C build -j${N}
        INSTALL_COMMAND ""
        BUILD_IN_SOURCE 1
      )
```

**Comment by beniz** [comment]
This is fixed with #1369, disabling MMAP, no alternative solution found at the moment :/ Thanks for the report.

**Comment by dgtlmoon** [comment]
@beniz no probs, this will make its way into a docker build at some time?
